cep-00??.md

```diff
@@ -0,0 +1,111 @@
+
+For each entry in the contents table, compute the hash for the concatenation of:
+- UTF-8 encoded bytes of the path, relative to the input directory. Backslashes MUST be normalized to forward slashes before encoding.
+- Then, depending on the type:
+    - For text files, the UTF-8 encoded bytes of an `F` separator, followed by the UTF-8 encoded bytes of its line-ending-normalized contents (`\r\n` replaced with `\n`). A file is considered
+    a text file if all the contents can be UTF-8 decoded. Otherwise it's considered binary. If the
+    file can't be opened, it's handled as if it were empty.
+    - For binary files, the UTF-8 encoded bytes of an `F` separator, followed by the bytes of its contents.
+    - For a directory, the UTF-8 encoded bytes of a `D` separator, and nothing else.
+    - For a symlink, the UTF-8 encoded bytes of an `L` separator, followed by the UTF-8 encoded bytes of the path it points to. Backslashes MUST be normalized to forward slashes before encoding.
+    - For any other paths, the UTF-8 encoded bytes of a `?` separator, and nothing else.
+- UTF-8 encoded bytes of the string `-`.
+
+Example implementation in Python:
+
+```python
+import hashlib
+from pathlib import Path
+
+def contents_hash(directory: str, algorithm: str) -> str:
+    hasher = hashlib.new(algorithm)
+    for path in sorted(Path(directory).rglob("*")):
+        hasher.update(path.relative_to(directory).replace("\\", "/").encode("utf-8"))
+        if path.is_symlink():
+            hasher.update(b"L")
+            hasher.update(str(path.readlink(path)).replace("\\", "/").encode("utf-8"))
+        elif path.is_dir():
+            hasher.update(b"D")
+        elif path.is_file():
+            hasher.update(b"F") 
+            try:
+                # assume it's text
+                lines = []
+                with open(path) as fh:
+                    for line in fh:
+                        lines.append(line.replace("\r\n", "\n")
+                for line in lines:
+                    hasher.update(line.encode("utf-8")))
+            except UnicodeDecodeError:
+                # file must be binary
+                with open(path, "rb") as fh:
+                    for chunk in iter(partial(fh.read, 8192), b""):
+                        hasher.update(chunk)
+        else:
+            hasher.update(b"?")
+        hasher.update(b"-")
+    return hasher.hexdigest()
+```
+
+## Motivation
+
+Build tools like `conda-build` and `rattler-build` need to fetch the source of the project being packaged. The integrity of the download is checked by comparing its known hash (usually SHA256) against the obtained file. If they don't match, an error is raised.
```

Comment:
> If they don't match, an error is raised.

This doesn't say what build tools should do when the files can't be read or an error is raised during computation, and the implementation in https://github.com/conda/conda-build/pull/5277/files#r1861269213 shows that it ignores file read errors (which can have many reasons) and only creates the content hash with the name of the file. I'd think that's a failure state for a content hash algorithm that aims to be consistent since it would expose it to file permission attacks etc.

Reply:
I'm not sure I follow how this could be weaponized as an attack vector. Let's say we have a directory with three files with `(name, content)`:

- `('file1.txt', '123')`
- `('file2.txt', '456')`
- `('file3.txt', '789')`

We end up computing the hash of these Utf-8 bytes:

```
file1.txtF123-file2.txtF456-file3.txtF789-
```
```python
>>> import hashlib
>>> hashlib.md5(b"file1.txtF123-file2.txtF456-file3.txtF789-").hexdigest()
'54866bc311f08b2e082466b090cbe560'
```

If let's say `file1.txt` changes permissions and becomes unreadable, then the string would be:

```
file1.txtF-file2.txtF456-file3.txtF789-
```

, which has a different hash.

If `file1.txt` becomes a directory it would then be:

```
file1.txtD-file2.txtF456-file3.txtF789-
```

Same for an unknown file type (but `?` instead of `D`).

Reply:
Or maybe we do need to error out, because an unreadable file might change contents arbitrarily and its hash would never change.